### PR TITLE
Add comprehensive docstring validation tool with enhanced RST error handling

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This PR adds a comprehensive docstring validation tool that uses Sphinx's parsing pipeline to catch RST formatting errors early in development. The tool helps prevent docstring issues from reaching the heavyweight Sphinx build process by providing fast, targeted validation.

**Key features:**
- Single symbol validation for quick development feedback
- Batch validation mode for entire modules (`--all-public` flag)
- Enhanced error handling with detailed RST parsing diagnostics
- CLI interface with verbose output options
- Integration with existing automation infrastructure

**Motivation:** Docstring parsing errors often only surface during full documentation builds, making them expensive to debug. This tool provides immediate feedback during development, improving documentation quality and reducing build failures.

## How I Tested These Changes

- Added comprehensive test suite covering common RST errors, valid symbol validation, and edge cases
- Tested CLI interface with various symbol paths and validation modes
- Validated against known good docstrings in the Dagster codebase
- Confirmed integration with automation package structure

```
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script) python scripts/validate_docstrings.py dagster.asset
Validating docstring for: dagster.asset
✓ Docstring is valid!
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script)
```

Inject error:

```
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script) python scripts/validate_docstrings.py dagster.asset
Validating docstring for: dagster.asset

ERRORS:
  - RST syntax: <string>:34: (ERROR/3) Unexpected indentation.
  - RST syntax: <string>:37: (ERROR/3) Unexpected indentation.
  - RST syntax: <string>:67: (ERROR/3) Unexpected indentation.
  - RST syntax: <string>:70: (ERROR/3) Unexpected indentation.
  - RST syntax: <string>:76: (ERROR/3) Unexpected indentation.
  - RST syntax: <string>:82: (ERROR/3) Unexpected indentation.

WARNINGS:
  - RST syntax: <string>:32: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
  - RST syntax: <string>:35: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
  - RST syntax: <string>:39: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
  - RST syntax: <string>:65: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
  - RST syntax: <string>:68: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
  - RST syntax: <string>:72: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
  - RST syntax: <string>:74: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
  - RST syntax: <string>:78: (WARNING/2) Block quote ends without a blank line; unexpected unindent.
  - RST syntax: <string>:80: (WARNING/2) Definition list ends without a blank line; unexpected unindent.
✗ Docstring validation failed
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script) ✗
```

Perf is .38 s wall:

```
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script) time  python scripts/validate_docstrings.py dagster.asset
Validating docstring for: dagster.asset
✓ Docstring is valid!
python scripts/validate_docstrings.py dagster.asset  0.38s user 0.07s system 99% cpu 0.451 total
(dagster) ➜  dagster git:(schrockn/sphinx-parse-script)
```

### CODE REVIEW

**Comment by schrockn** [comment]
* **#31283** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31283?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31282** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31282?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#31281** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31281?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31281?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#31280** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/31280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/automation/automation/docstring_lint/cli.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    """
+    # Core use case - validate single docstring efficiently
+    validator = DocstringValidator()
+
+    try:
+        if all_public:
+            # Batch validation mode
+            importer = SymbolImporter()
+            symbols = importer.get_all_public_symbols(symbol_path)
+            print(f"Validating {len(symbols)} public symbols in {symbol_path}\n")  # noqa: T201
```
Comment (schrockn) [inline python_modules/automation/automation/docstring_lint/cli.py:31]: These get converted to echo in https://app.graphite.dev/github/pr/dagster-io/dagster/31299/Use-click.echo-instead-of-print

File: python_modules/automation/automation/docstring_lint/validator.py [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+# Filter out Sphinx role errors and warnings that are valid in the full Sphinx build context
+# but appear as errors in standalone docutils validation. These roles like :py:class:,
+# :func:, etc. are properly resolved during the actual documentation build with all
+# Sphinx extensions loaded, but docutils alone doesn't recognize them.
+# Set to False to see all role-related errors/warnings for debugging.
+DO_FILTER_OUT_SPHINX_ROLE_WARNINGS = True
+
+# Setup paths to match Dagster's Sphinx configuration
+DAGSTER_ROOT = Path(__file__).parent.parent.parent.parent.parent
+SPHINX_EXT_PATH = DAGSTER_ROOT / "docs" / "sphinx" / "_ext"
```
Comment (cmpadden) [inline python_modules/automation/automation/docstring_lint/validator.py:26]: Nice, just wanted to confirm that we're loading our custom extensions.

**Comment by schrockn** [comment]
Yeah my thinking is that this a pretty isolated toolchain and we have a backstop with the old workflow. We can remove this if it becomes not worth it or if there is a successor toolchain (🤞🏻 for astral to save us in this domain as well)

**Comment by schrockn** [comment]
### Merge activity

* **Jul 23, 7:59 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31281).
* **Jul 23, 8:03 PM UTC**: [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31281) rebased this pull request as part of a merge.
* **Jul 23, 8:04 PM UTC**: @schrockn merged this pull request with [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/31281).
